### PR TITLE
feat: More controlled deferred revenue booking

### PR DESCRIPTION
**Pull request by deepeshgarg007**
image

- Provision to book deferred revenue via Journal Entry
- Provision to save Journal Entry as draft for reviewing and submit
- Provision to book deferred revenue/expense on monthly basis rather than by days

### CODE REVIEW

File: erpnext/accounts/deferred_revenue.py [hunk 1]
```diff
@@ -152,12 +165,24 @@ def calculate_amount(doc, item, last_gl_entry, total_days, total_booking_days, a
...
+		journal_entry_details = frappe.db.sql('''
+			SELECT sum(c.{0}) as total_credit, sum(c.{1}) as total_credit_in_account_currency, reference_detail_no
+			FROM `tabJournal Entry` p , `tabJournal Entry Account` c WHERE p.name = c.parent and
+			p.company = %s and c.account=%s and c.reference_type=%s and c.reference_name=%s and c.reference_detail_no=%s
+			and p.docstatus < 2 group by reference_detail_no
+		'''.format(total_credit_debit, total_credit_debit_currency),
+			(doc.company, item.get(deferred_account), doc.doctype, doc.name, item.name), as_dict=True)
+
 		already_booked_amount = gl_entries_details[0].total_credit if gl_entries_details else 0
+		already_booked_amount += journal_entry_details[0].total_credit if journal_entry_details else 0
```
Comment (nabinhait) [inline erpnext/accounts/deferred_revenue.py]: same entry will be added twice via GL and JE?

Comment (nabinhait) [reply erpnext/accounts/deferred_revenue.py]: I think it should only consider draft JE

File: erpnext/accounts/deferred_revenue.py [hunk 2]
```diff
@@ -166,7 +191,7 @@ def book_deferred_income_or_expense(doc, deferred_process, posting_date=None):
 	enable_check = "enable_deferred_revenue" \
 		if doc.doctype=="Sales Invoice" else "enable_deferred_expense"
 
-	def _book_deferred_revenue_or_expense(item):
+	def _book_deferred_revenue_or_expense(item, via_je, submit_je):
```
Comment (nabinhait) [inline erpnext/accounts/deferred_revenue.py]: use `via _journal_entry` and `submit_journal_entry`

File: erpnext/accounts/doctype/journal_entry/journal_entry.json [hunk 3]
```diff
@@ -83,7 +83,7 @@
    "label": "Entry Type",
    "oldfieldname": "voucher_type",
    "oldfieldtype": "Select",
-   "options": "Journal Entry\nInter Company Journal Entry\nBank Entry\nCash Entry\nCredit Card Entry\nDebit Note\nCredit Note\nContra Entry\nExcise Entry\nWrite Off Entry\nOpening Entry\nDepreciation Entry\nExchange Rate Revaluation",
+   "options": "Journal Entry\nInter Company Journal Entry\nBank Entry\nCash Entry\nCredit Card Entry\nDebit Note\nCredit Note\nContra Entry\nExcise Entry\nWrite Off Entry\nOpening Entry\nDepreciation Entry\nExchange Rate Revaluation\nDeferred Revenue Entry\nDeferred Expense Entry",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/journal_entry/journal_entry.json]: Use simple "Deferred Revenue" and "Deferred Expense"

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/accounts/deferred_revenue.py (hunk 1 above)
Comment (deepeshgarg007) [reply erpnext/accounts/deferred_revenue.py]: > same entry will be added twice via GL and JE?

@nabinhait 
While getting the already booked amount via GLE , the GLE's having voucher type only as 'Sales Invoice or Purchase Invoice' will be considered so same entry should not be added twice. Entries booked via Journal Entries will not be considered. Thats why I am getting both Draft and Submitted entries while fetching the amount from Journal Entry

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.json [hunk 4]
```diff
@@ -189,13 +194,47 @@
...
+  },
+  {
+   "fieldname": "column_break_18",
+   "fieldtype": "Column Break"
+  },
+  {
+   "default": "No",
+   "description": "If this is set as 'No' direct GL Entries will be created to book Deferred Revenue/Expense",
+   "fieldname": "book_deferred_entries_via_journal_entry",
+   "fieldtype": "Select",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.json]: Why not checkbox?

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.json [hunk 5]
```diff
@@ -189,13 +194,47 @@
...
+   "label": "Book Deferred Entries Via Journal Entry",
+   "options": "No\nYes"
+  },
+  {
+   "default": "No",
+   "depends_on": "eval:doc.book_deferred_entries_via_journal_entry == 'Yes'",
+   "description": "If this is set as 'No' Journal Entries will be saved in a Draft state and will have to be submitted manuallly",
+   "fieldname": "submit_journal_entries",
+   "fieldtype": "Select",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.json]: Why not checkbox?

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Review by nabinhait — APPROVED** [review]

**Comment by nabinhait** [comment]
@deepeshgarg007 Update docs

**Pull request merged**